### PR TITLE
Core: Fix lazy snapshot loading history

### PR DESCRIPTION
**Pull request by rdblue**
This is a fix for table history when snapshots are lazily loaded (introduced by #6811).

When table metadata is returned with some snapshots suppressed, the next time table metadata is rebuilt results in losing history because the snapshot log is rewritten and removes any history that references unknown snapshots. The solution is to only clear history for unknown snapshots if there is a `RemoveSnapshot` change.

In addition, this PR adds a method to suppress snapshots that are not directly referenced by a ref for services that lazily load snapshots. This avoids the same history problem on the server side.

Closes #8096.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 1]
```diff
@@ -1253,20 +1253,53 @@ public Builder removeStatistics(long snapshotId) {
...
+    /**
+     * Rewrite this builder's snapshots by removing the snapshots for a list of IDs.
+     *
+     * <p>If suppress is true, changes are not created.
+     *
+     * @param idsToRemove collection of snapshot IDs to remove from this builder
+     * @param suppress whether the operation is suppressing snapshots (retains history) or removing
+     * @return this for method chaining
+     */
+    public Builder rewriteSnapshotsInternal(Collection<Long> idsToRemove, boolean suppress) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:1293]: did you mean to make this one `public`?

**Review by nastra — COMMENTED** [review]
LGTM but I added one comment around the visibility of a method. I've also ported the original test I had via https://github.com/rdblue/iceberg/pull/58

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1293]: No, good catch!

**Review by danielcweeks — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks for reviewing, @nastra and @danielcweeks!

**Comment by rdblue** [comment]
@nastra, sorry I missed that you'd added a PR against mine with a test. Can you open a new PR with it?

**Comment by nastra** [comment]
@rdblue I've opened https://github.com/apache/iceberg/pull/8489 to add the test

**Pull request merged**